### PR TITLE
feat: Delegated routing with custom configuration.

### PR DESCRIPTION
**Pull request by ajnavarro**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,5 +1,7 @@
 module github.com/ipfs/kubo
 
+replace github.com/libp2p/go-libp2p-routing-helpers => /home/ajnavarro/workspace/go-libp2p-routing-helpers
```
Comment (ajnavarro) [inline go.mod:3]: TODO: remove

File: test/sharness/t0701-delegated-routing-reframe.sh [hunk 2]
```diff
@@ -67,12 +67,31 @@ test_expect_success "no routers means findprovs returns no results" '
...
+      },
+      "provide": {
+        "RouterName": "TestDelegatedRouter"
+      },
+      "put-ipns": {
+        "RouterName": "TestDelegatedRouter"
+      }
+    }' || exit 1
 
 test_expect_success "adding reframe endpoint to Routing.Routers config works" '
```
Comment (ajnavarro) [inline test/sharness/t0701-delegated-routing-reframe.sh:96]: Add a smoke test for errors (example: check number of methods or duplicate methods on config)

**Review by guseggert — CHANGES_REQUESTED** [review]

File: routing/delegated.go [hunk 3]
```diff
@@ -1,93 +1,302 @@
...
-
-var _ TieredRouter = &Tiered{}
+var log = logging.Logger("routing/delegated")
 
-// Tiered is a routing Tiered implementation providing some extra methods to fill
-// some special use cases when initializing the client.
-type Tiered struct {
-	routinghelpers.Tiered
-}
+func Parse(routers config.Routers, methods config.Methods, extraDHT *ExtraDHTParams, extraReframe *ExtraReframeParams) (routing.Routing, error) {
```
[the remark this answers is not in the document]
Comment (ajnavarro) [reply routing/delegated.go:29]: WDYT about creating a fix for this after making the RC?

**Comment by BigLep** [comment]
@ajnavarro : I don't think we should say "closes https://github.com/ipfs/kubo/issues/9150" since that issue is done when reframe is used in Kubo in production.  This PR is an important step but not the end.

**Comment by ajnavarro** [comment]
@BigLep removed that issue from the list.

**Review by guseggert — APPROVED** [review]
